terraform/evaluate.go

```diff
@@ -935,3 +945,44 @@ func moduleDisplayAddr(addr addrs.ModuleInstance) string {
 		return addr.String()
 	}
 }
+
+// markProviderSensitiveAttributes returns an updated value
+// where attributes that are Sensitive are marked
+func markProviderSensitiveAttributes(schema *configschema.Block, val cty.Value) cty.Value {
+	return val.MarkWithPaths(getValMarks(schema, val, nil))
+}
+
+func getValMarks(schema *configschema.Block, val cty.Value, path cty.Path) []cty.PathValueMarks {
+	var pvm []cty.PathValueMarks
+	for name, attrS := range schema.Attributes {
+		if attrS.Sensitive {
+			attrPath := append(path, cty.GetAttrStep{Name: name})
```

Comment:
I think this (and the other use of `append(path, …)` below is subject to the same path array reuse bug I fixed in https://github.com/zclconf/go-cty/pull/70. Because we're storing these paths, I think we need to copy them explicitly here instead of relying on `append` to reallocate.

Reply:
This was a really good thing to flag, but I _think_ this is okay, and I'll explain why: in this case, I'm allocating a new value `attrPath`, not mutating `path` in place, which was the source of the other bug. This should be a separate value, no?

And the one below ... I don't think the same path issue applies, as this is not a Path but a slice of PathValueMarks, so I _think_ this append mutation is alright.

Reply:
The slice is a new allocation, but the backing array might not be, if it's already large enough such that `append` doesn't force a reallocation. [Here's a playground example which only assigns each value once and exhibits the bug](https://play.golang.org/p/ahBFd_p_0lA). Basically, when an `append` on a slice of (`len=2`, `cap=2`) causes the backing array to be reallocated to `cap=4`, subsequent `append`s inside a loop will overwrite each other.

I'm not completely sure that the bug is present in this function, as its recursiveness is a bit difficult for me to follow. I'm having trouble thinking about how to write a test case to verify either way. I think we'd need deeply nested blocks (n=3 or 4?) where each has multiple sensitive attributes.

Reply:
New commit to do the copying: https://github.com/hashicorp/terraform/pull/26590/commits/394e60608ccee16c4c979b5843a614bbbb856f63 I did add a case of nested block here too (although it didn't surface the suspected bug).